src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4740,3 +4740,101 @@ void C2_MacroAssembler::udivmodL(Register rax, Register divisor, Register rdx, R
 }
 #endif
 
+void C2_MacroAssembler::float_class_check(int opcode, Register dst, XMMRegister src, Register temp) {
+
+  int32_t POS_INF = 0x7f800000;
+  int32_t KILL_SIGN_MASK = 0x7fffffff;
+
+  movdl(temp, src);
+  andl(temp, KILL_SIGN_MASK);
+  cmpl(temp, POS_INF);
+  switch (opcode) {
+    case Op_IsFiniteF:
+      setb(Assembler::below, dst);
+      break;
+    case Op_IsInfiniteF: {
+      setb(Assembler::equal, dst);
+      break;
+    }
+    case Op_IsNaNF:
+       setb(Assembler::above, dst);
+       break;
+    default:
+      assert(false, "%s", NodeClassNames[opcode]);
+  }
+}
+
+void C2_MacroAssembler::float_class_check_vfp(int opcode, Register dst, XMMRegister src, KRegister tmp) {
+  uint8_t imm8;
+  switch (opcode) {
+    case Op_IsFiniteF:
+      imm8 = 0x99;
+      break;
+    case Op_IsInfiniteF:
+      imm8 = 0x18;
+      break;
+    case Op_IsNaNF:
+       imm8 = 0x81;
+       break;
+    default:
+      assert(false, "%s", NodeClassNames[opcode]);
+  }
+  vfpclassss(tmp, src, imm8);
+  kmovbl(dst, tmp);
+  if (opcode == Op_IsFiniteF) {
+    xorl(dst, 0x00000001);
```

Comment:
`notl(dst)`?

Reply:
I can't recall why `notl(dst)` didn't work. Will try it and let you know.

Reply:
notl(dst) doesn't work. we need to flip only the last bit as the kreg stores the result in the last bit.